### DISCUSSION TITLE
Is it possible to target WebGPU rather than WebGL when building for wasm?

### DISCUSSION THREAD

**Discussion by spencerkohan** [gfx-rs/wgpu#3118] [Q&A]
I've managed to get a project building for the web, but it seems to be targeting the WebGL back-end.  Is it possible to target WebGPU on browsers which support it, like Firefox Nightly?

If so, how would I go about doing this?

**Comment by grovesNL** [comment]
Hi! 👋 You can target WebGL or WebGPU by following the guide in the wiki for this repository: https://github.com/gfx-rs/wgpu/wiki/Running-on-the-Web-with-WebGPU-and-WebGL

**Reply by spencerkohan** [reply]
Thanks for the quick response!

How does the WebGPU/WebGL back-end decision actually get made?  I am trying to run an example based on the code in the learn-wgpu tutorials, and running it with the method described in the document link still appears to be going through WebGL (I think).

I.e. I see this error when I try to load:

> Error in Adapter::request_device: Limit 'max_compute_workgroups_per_dimension' value 65535 is better than allowed 0

I believe that would be a limit imposed by the webgl downlevel limits no?

**Reply by grovesNL** [reply]
It's selected at compile-time (by the presence of the `webgl` feature) for now, but we plan to allow run-time selection later.

> I believe that would be a limit imposed by the webgl downlevel limits no?

That's right - compute shaders aren't supported by WebGL.

**Reply by spencerkohan** [reply]
Ok that explains it - I had the feature "webgl" declared in my Cargo.toml.

But if I remove it I get all these errors when compiling for the wasm target:

```
error[E0432]: unresolved import `web_sys::GpuTextureFormat`
   --> /Users/spencerkohan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.13.1/src/backend/web.rs:511:9
    |
511 |     use web_sys::GpuTextureFormat as tf;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `GpuTextureFormat` in the root

error[E0432]: unresolved import `web_sys::GpuTextureSampleType`
   --> /Users/spencerkohan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.13.1/src/backend/web.rs:561:9
    |
561 |     use web_sys::GpuTextureSampleType as ts;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `GpuTextureSampleType` in the root

error[E0432]: unresolved import `web_sys::GpuCullMode`
   --> /Users/spencerkohan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.13.1/src/backend/web.rs:573:9
    |
573 |     use web_sys::GpuCullMode as cm;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ no `GpuCullMode` in the root

error[E0432]: unresolved import `web_sys::GpuFrontFace`
   --> /Users/spencerkohan/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.13.1/src/backend/web.rs:583:9
    |
583 |     use web_sys::GpuFrontFace as ff;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `GpuFrontFace` in the root
...
```

Do you know what I might be missing?

**Reply by grovesNL** [reply]
You need to specify the `RUSTFLAGS` environment variable to opt-in to the unstable WebGPU API. There's an example under the "wasm-bindgen-cli" section of the wiki at https://github.com/gfx-rs/wgpu/wiki/Running-on-the-Web-with-WebGPU-and-WebGL#manual-compilation-with-wasm-bindgen-cli
